### ISSUE TITLE
Reduce max length of signature in vote message

### ISSUE THREAD

**Issue by djrtwo** [ethereum/casper#110]
# Issue
Vote messages must be less that or equal 1024 bytes, defined by the type in the `vote` method signature. When parsing the signature of the vote from the vote message, we currently only enforce that the signature too is less than or equal to 1024 bytes.

https://github.com/ethereum/casper/blob/d5aed932634e28ab90d31facc6d9944be1a5d426/casper/contracts/simple_casper.v.py#L402

Due to the variable amount of bytes required to encode the other elements in the list, there is a range on the maximum length of a signature depending on the epoch or even the `validator_index`. To enforce more strict requirements, we propose restricting `sig` to length less than or equal to 934 bytes. This number assumes the other elements of the vote message take their maximal length to encode.

1024 bytes available
3 to encode the whole list
17 worst case to encode an int128
33 to encode the bytes32 hash
3 to encode the signature bytes

1024 - 3 - 17*3 - 33 - 3 == 934 bytes max for signature

Sanity checked with the following python code
```python
import rlp
validator_index = target_epoch = source_epoch = 2**128 - 1
sig = b'\xff' * 934
target_hash = b'\xff' * 32
len(rlp.encode([validator_index, target_hash, target_epoch, source_epoch, sig])) == 1024
```
Note, the `logout` message has fewer elements so the signature could theoretically be larger than `934` for this action, but to reduce complexity, 934 should be used for `logout` messages as well.

# Proposed Implementation
- Define MAX_SIGNATURE_LENGTH as 934
- Enforce sig as `<= MAX_SIGNATURE_LENGTH` in `vote`, `slash`, and `logout`
- tests.

**Comment by hrishikeshio** [comment]
@djrtwo Would it be better to add this enforcement in validate_signature()?

**Comment by djrtwo** [comment]
The type declaration should handle the enforcement. I was thinking we just make the type declaration the following anywhere we are referencing the sig which is a number of places at this point (including `validate_signature`)

```python
sig: bytes <= MAX_SIGNATURE_LENGTH = ...
```

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 200.0 DAI (200.0 USD @ $1.0/DAI) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/casper/110/391).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $8,316.90 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.

 1. [@iamonuwa](https://gitcoin.co/profile/iamonuwa) 

 has committed to working on this project to be completed 5 months, 4 weeks from now.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 7 months, 2 weeks ago.
Please review their action plans below:


**1) [mkeen](https://gitcoin.co/profile/mkeen) has started work.**

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/casper/110/391).

**Comment by mkeen** [comment]
@djrtwo Your proposed solution is syntactically elegant, but from what I can see, it's going to require an enhancement to Vyper. Vyper's `RLPList` function [sets the size of 
all byte array members equal to the size of the entire RLPList](https://github.com/ethereum/vyper/blob/0d92d86752bcfca875e042cec6048488db3d479c/vyper/functions/functions.py#L431).

For the purposes of decoding incoming data, Vyper's `RLPList` requires as arguments a bytearray, and a serial list of types -- the latter is used by RLPList to decode the former into an array of native objects. It is used like this by Casper:

https://github.com/ethereum/casper/blob/dcf4caf35937800b04c6bac4a2ab13c1513b6147/casper/contracts/simple_casper.v.py#L491

The above doesn't expose any way to let Vyper's `RLPList` know how many bytes will be in the byte array members, so (again, linked above) `RLPList` just allocates a number of bytes equal to the length of the entire message, in our case, `1024`.

To summarize, due to the current implementation of `RLPList` in Vyper, the size of the variable `sig` referenced below must be equal to the size of the entire message (`vote_msg`) it's contained in.

https://github.com/ethereum/casper/blob/dcf4caf35937800b04c6bac4a2ab13c1513b6147/casper/contracts/simple_casper.v.py#L489-L496

I've looked into making some enhancements to the Vyper `RLPList` implementation, but it's a fairly involved undertaking, and I'd need to understand more about related project [pyrlp](https://github.com/ethereum/pyrlp). I may continue down this rabbit hole, but for now, this should probably be put on hold.

Requesting a suggestion on how I should proceed.

**Comment by gitcoinbot** [comment]
@mkeen Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=100)

**Comment by mkeen** [comment]
I am still working on this. Submitted a proposal to make this possible with a change to Vyper.

**Comment by gitcoinbot** [comment]
@mkeen Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=100)

**Comment by mkeen** [comment]
@djrtwo Can I get a 10 day snooze on this? RLPList is undergoing heavy changes RN.

**Comment by djrtwo** [comment]
@mkeen I can't snooze as I'm not the founder. will have to ask @gdipri01

**Comment by djrtwo** [comment]
Cool I just saw that the VIP got approved. Congrats!

**Comment by mkeen** [comment]
Thanks!

**Comment by vs77bb** [comment]
@mkeen @djrtwo Just 10-day-snoozed Gitcoin Bot, appreciate the request and patience as we work out the kinks 🙂

**Comment by weberV** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 200.0 DAI (200.0 USD @ $1.0/DAI) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/ethereum/casper/110/391).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $11,355.67 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by vs77bb** [comment]
Hi @mkeen just checking in, is this one still pending? Hope you are doing well 🙂

**Comment by djrtwo** [comment]
This is blocked by the implementation of a VIP https://github.com/ethereum/vyper/issues/818

It appears to be in their "final countdown" before beta. So I expect it to be finished in the relative short term

**Comment by gitcoinbot** [comment]
@mkeen Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] warning (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=100)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

@mkeen due to inactivity, we have escalated [this issue](https://gitcoin.co/issue/ethereum/casper/110/391) to Gitcoin's moderation team. Let us know if you believe this has been done in error!

* [x] warning (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=1) | [3 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=3) | [5 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=5) | [10 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=10) | [100 days](https://gitcoin.co/issue/ethereum/casper/110/391?snooze=100)

**Comment by mkeen** [comment]
Hi @vs77bb yes still pending! But I am keeping an eye on it.

**Comment by kuhnchris** [comment]
Hi there everyone. Is this issue still blocked due to your dependencies? 
Let me know if we can help you in any ways!

Thanks!

**Comment by hrishikeshio** [comment]
@kuhnchris The repo is deprecated. No commits since Jun 2018.

**Comment by kuhnchris** [comment]
Well, there are still 200$ in DAI stashed in the bounty from Greg. If this repository is deprecated he should cancel the bounty so he can at least get his funds back. We'll try to mail him, thanks for the info!
May I note that if this is deprecated that maybe the repository should be changed to read-only, respectively archived?

Thanks!
Chris

**Comment by mkeen** [comment]
The repo isn't deprecated as far as I know. It would be nice to at least partially claim the bounty, since I did a bunch of analysis and spec'ed out a fix upstream. Just a thought :)

**Comment by kuhnchris** [comment]
@djrtwo @hrishikeshio your calls. cc @vs77bb

**Comment by mkeen** [comment]
@djrtwo @hrishikeshio @vs77bb LMK. I think it's only fair.

**Comment by hrishikeshio** [comment]
Its deprecated according to https://github.com/ethereum/casper/blob/master/README.md.
Anyways, I am not the decision maker or bounty sponsorer. So its not my decision. good luck.

**Comment by mkeen** [comment]
Yikes.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. **Submitted** 4. Done 

<hr>

__Work for 200.0 DAI (200.0 USD @ $1.0/DAI) has been submitted by__:

 1. [@mkeen](https://gitcoin.co/profile/mkeen)



@gdipri01 please take a look at the submitted work:
* (Link Not Provided) by @mkeen

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ethereum/casper/110/391)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ethereum/casper/110/391).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $86,212.92 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
